windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -988,98 +1010,54 @@ procedure InstallingState.HandleInstallNow;
   // Do Nothing. Need the UI to let user know installation in progress OR
 end;
 
-{ RetryState }
-
-procedure RetryState.Enter;
+procedure InstallingState.HandleFirstRun;
 begin
-  bucStateContext.SetRegistryState(usRetry);
+  bucStateContext.HandleMSIInstallComplete;
+  // Result := kmShellContinue;
 end;
 
-procedure RetryState.Exit;
-begin
-
-end;
-
-procedure RetryState.HandleCheck;
-begin
-
-end;
-
-function RetryState.HandleKmShell;
-begin
-  // #TODO: #10210 Implement retry
-  Result := kmShellContinue
-end;
-
-procedure RetryState.HandleDownload;
-begin
-
-end;
-
-procedure RetryState.HandleAbort;
-begin
-
-end;
-
-procedure RetryState.HandleInstallNow;
-begin
-  // TODO: #10038 handle retry counts
-  ChangeState(InstallingState);
-end;
-
-{ PostInstallState }
-
-procedure PostInstallState.Enter;
-begin
-  // Enter downloading state
-  bucStateContext.SetRegistryState(usPostInstall);
-end;
-
-procedure PostInstallState.Exit;
-begin
-
-end;
-
-procedure PostInstallState.HandleCheck;
-begin
-  // Handle Check
-end;
-
-function PostInstallState.HandleKmShell;
-begin
-  HandleMSIInstallComplete;
-  Result := kmShellContinue;
-end;
-
-procedure PostInstallState.HandleDownload;
-begin
-  // Do Nothing
-end;
-
-procedure PostInstallState.HandleMSIInstallComplete;
+// Private Functions:
+function ConfigCheckContinue: Boolean;
 var
-  SavePath: string;
-  fileName: String;
-  Filenames: TStringDynArray;
+  Registry: TRegistryErrorControlled;
 begin
-  SavePath := IncludeTrailingPathDelimiter(TKeymanPaths.KeymanUpdateCachePath);
-
-  GetFileNamesInDirectory(SavePath, Filenames);
-  for fileName in Filenames do
-  begin
-    System.SysUtils.DeleteFile(fileName);
+  { Verify that it has been at least CheckPeriod days since last update check }
+  Result := False;
+  try
+    Registry := TRegistryErrorControlled.Create; // I2890
+    try
+      if Registry.OpenKeyReadOnly(SRegKey_KeymanDesktop_CU) then
+      begin
+        if Registry.ValueExists(SRegValue_CheckForUpdates) and
+          not Registry.ReadBool(SRegValue_CheckForUpdates) then
+        begin
+          Result := False;
+          Exit;
+        end;
+        if Registry.ValueExists(SRegValue_LastUpdateCheckTime) and
+          (Now - Registry.ReadDateTime(SRegValue_LastUpdateCheckTime) >
+          CheckPeriod) then
+        begin
+          Result := True;
+        end
+        else
+        begin
+          Result := False;
+        end;
+        Exit;
+      end;
+    finally
+      Registry.Free;
+    end;
+  except
+    { we will not run the check if an error occurs reading the settings }
+    on E: Exception do
+    begin
+      Result := False;
+      LogMessage(E.Message);
```

Comment:
Is this the right thing to call? Or do we have to call `TKeymanSentryClient.Client.MessageEvent` here?